woocommerce/woocommerce-gutenberg-products-block
# Test Google's "ndb" E2E testing tool

### Notes

This PR showcases how we can use the "[ndb](https://github.com/GoogleChromeLabs/ndb)" debugging tool (i.e. Chromium DevTools) to debug our E2E test files.

### Testing

#### Automated Tests

### Manual Testing

1. Check out this PR.
2. Start Docker Desktop.
3. Run `npm run wp-env destroy && npm run wp-env start` to start a fresh e2e environment.
4. Run `npm run debug` to run "[ndb](https://github.com/GoogleChromeLabs/ndb)"
5. Create a test example or use an existing one under: `tests/e2e/specs`
6. If you want add a script to `package.json` to run only a specific file in your test. Example:

```json
"test:e2e-dev-my-single-test": "npm run wp-env:config && cross-env NODE_CONFIG_DIR=tests/e2e/config wp-scripts test-e2e --config tests/e2e/config/jest.config.js --puppeteer-interactive -- tests/e2e/specs/shopper/my-test-example.test.js",
```

7. In the "ndb" Chromium window, search and select your test file with: `Ctrl`/`Cmd` + `P`, and past the name of your test file.

<img width="850" alt="image" src="https://user-images.githubusercontent.com/14235870/156786706-ab6b13a2-e633-4c79-91af-dabea67d18e0.png">

8. Add a breakpoint to the test file

<img width="588" alt="image" src="https://user-images.githubusercontent.com/14235870/156787999-0f6d1ad5-20ae-431a-a3c5-f7da593b6139.png">

9. In the same window, search for `test:e2e-dev-my-single-test` under the `NPM Scripts` section (if you've created added this script in an earlier step, otherwise select any existing scripts that you want to execute) and hit `Run`.

<img width="355" alt="image" src="https://user-images.githubusercontent.com/14235870/156787134-f9f559f3-660d-4714-a130-fb274ef2ed4c.png">

10. The test execution should stop at the breakpoint. You can now add more breakpoints, step to the next line, resume, step into the next function call, check the Call Stack, etc.

<img width="1737" alt="image" src="https://user-images.githubusercontent.com/14235870/156788729-7c042d0a-0793-4975-99c8-4e9aa46886de.png">

### Changes made

The only changes made in this PR are:

* [install ndb as a development dependency](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5997/commits/8db921a5d8e77872d1a93a1a8aed345bf597cedf)
* [Add "debug" script to package.json to launch "ndb"](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5997/commits/80906edffe7c11018e37010bf5e763ee70ee91e5)
